### DISCUSSION TITLE
[boost-log:x64-windows] possible port degradation from 1.75 to 1.78

### DISCUSSION THREAD

**Discussion by ssh4net** [microsoft/vcpkg#22762] [Q&A]
**Describe the bug**
Installed boost-log:x64-windows generate compilation error in VS:
Error	LNK2038	mismatch detected for 'boost_log_abi': value 'v2_mt_nt62' doesn't match value 'v2_mt_nt6' in ****.obj	****	***\ImageUtils.obj	1	

**Environment**
- OS: Windows 10 Pro
- Compiler: VS2019

**To Reproduce**
1. ./vcpkg install boost-log:x64-windows
2. try to compile existing project that use boost-log
2. See error

**Expected behavior**
v 1.75 do not have this issue.

**Additional context**
from 1.78 boost changelog:
On Windows, when building the library for Windows 8 or later, the library will use nt62 tag in the version namespace to denote the target OS ABI. For example, the version namespace could be named as v2_mt_nt62. This name will be part of all symbols exported by the library. Use the BOOST_USE_WINAPI_VERSION macro consistenly when building Boost and your code to request the minimum target Windows version.

Sound like boot-log port missed mandatory for correct build macro.

**Comment by PhoebeHui** [comment]
@ssh4net, thanks for reporting this issue! could you please share the repro code with us, or a project to help us repro this issue quickly?

**Comment by ssh4net** [comment]
Sorry, can't share code or project. And not skilled too much to make minimal possible from existing code that reproduce this bug.

But my coder gave me temporary fix for issue 
compiling project with this option:
`-DBOOST_USE_WINAPI_VERSION=BOOST_WINAPI_VERSION_WIN7`

looks like this is a bug in Boost itself

**Comment by MarcelNehring** [comment]
I saw this issue while looking for a solution for what looks like this exact problem.
This isn't really a **minimal** example but I hope it suffices to analyze the issue further.

Building this code fails for me as soon as I define `_WIN32_WINNT` to target Windows 8 or later.
As @ssh4net and the boost changelog already mentioned defining `BOOST_USE_WINAPI_VERSION` to `BOOST_WINAPI_VERSION_WIN7` also fixes the linker errors.

```C++
// Setting this to 0x0602 (Windows 8) or higher will cause linking to fail.
// It works with 0x0601 (Windows 7).
#define  _WIN32_WINNT   0x0602

// Uncomment this to fix linker errors independent of targeted Windows version.
//#define BOOST_USE_WINAPI_VERSION BOOST_WINAPI_VERSION_WIN7

#include <boost/log/sources/logger.hpp>
#include <boost/log/sources/record_ostream.hpp>
#include <boost/log/sources/global_logger_storage.hpp>
#include <boost/log/sources/severity_logger.hpp>

enum class LogLevel
{
    LEVEL_DEBUG,
    LEVEL_ERROR
};

BOOST_LOG_INLINE_GLOBAL_LOGGER_DEFAULT(MyLogger, boost::log::sources::severity_logger_mt<LogLevel>)

#define LOGGER(lvl) BOOST_LOG_SEV(MyLogger::get(), lvl)

int main()
{
    LOGGER(LogLevel::LEVEL_DEBUG) << "Hello World!\n";
}
```

**Comment by PhoebeHui** [comment]
@MarcelNehring, thanks for the example code! it seems Upstream by design issue, relate to https://github.com/boostorg/log/issues/172.

`BOOST_USE_WINAPI_VERSION` is defined to `BOOST_WINAPI_VERSION_WIN7` by default, and `_WIN32_WINNT` and `WINVER` are defined in boost-winapi header[1], which value is `BOOST_WINAPI_VERSION_WIN7`[2], that cause conflicts when user define these two macros in their custom project. Do you have any ideas to handle this case? cc @yurybura @Kojoley @strega-nil-ms @BillyONeal 

[1] https://github.com/boostorg/winapi/blob/develop/include/boost/winapi/config.hpp#L214-L219
```C++
#if !defined(_WIN32_WINNT)
#define _WIN32_WINNT BOOST_USE_WINAPI_VERSION
#endif
#if !defined(WINVER)
#define WINVER BOOST_USE_WINAPI_VERSION
#endif
```

[2] https://github.com/boostorg/winapi/blob/develop/include/boost/winapi/config.hpp#L105-L120
```C++
#if !defined(BOOST_USE_WINAPI_VERSION)
#if defined(_WIN32_WINNT)
#define BOOST_USE_WINAPI_VERSION _WIN32_WINNT
#elif defined(WINVER)
#define BOOST_USE_WINAPI_VERSION WINVER
#else
// By default use Windows 7 API on compilers that support it and Vista or XP on the others
#if (defined(_MSC_VER) && _MSC_VER < 1500) || defined(BOOST_WINAPI_IS_MINGW)
#define BOOST_USE_WINAPI_VERSION BOOST_WINAPI_VERSION_WINXP
#elif (defined(_MSC_VER) && _MSC_VER < 1600)
#define BOOST_USE_WINAPI_VERSION BOOST_WINAPI_VERSION_WIN6
#else
#define BOOST_USE_WINAPI_VERSION BOOST_WINAPI_VERSION_WIN7
#endif
#endif
#endif
```

**Comment by Kojoley** [comment]
Do I understand correctly that you install boost-log via vcpkg, then in your code that uses boost-log you define `_WIN32_WINNT` to some value (and which differs from one boost-log would guess, or you had supplied to vcpkg explicitly) - you are getting a linkage error?

**Comment by MarcelNehring** [comment]
@Kojoley yes that is correct.

I have not tried it yet but you could probably install boost-log while setting `_WIN32_WINNT` to the desired value using `VCPKG_CXX_FLAGS` in a custom triplet and thus get a matching boost-log version that does not produce linker errors.

**Comment by Kojoley** [comment]
Obviously vcpkg couldn't know which Windows version you want to target if you haven't tell that to it.
@PhoebeHui maybe a warning could be added when boost-log is installed telling that it will be built using a default `_WIN32_WINNT` value?

**Comment by BillyONeal** [comment]
I don't think we need a specific-to-vcpkg answer here. Triplets already let you set `_WIN32_WINNT`.

**Comment by johnco3** [comment]
I also have the above linkage issue.  Boost-Log frequently has porting issues that could easily be flushed out prior to pushing the port - I remember having to manually patch boost headers a few releases ago when I used to download the binaries & prebuilt headers from the boost.org website.  This was prior to making the switch to building boost using vcpkg & triplets - (in my case x64-windows).  I thought this would make upgrading to the latest versions of boost (in particular boost-log) a bit easier - unfortunately I was wrong.  I was able to work around the issue by setting a common preprocessor setting _WIN32_WINNT=0x0601 for all of my projects - not just the executable that performs the linkage.   I would prefer not to force the older Windows 7 API - but there you go.  I wish I knew how to make a local git patch to work around this problem - now I have to remember to remove the _WIN32_WINNT=0x0601 at some later date when boost finally gets around to fixing the issue.

**Comment by chrullrich** [comment]
To use Boost.Log in an application that is compiled with `_WIN32_WINNT` >= 0x0602, Boost.Log itself must be built that way. However, this also does not work.

Instead, the Boost.Log build fails with missing symbols, `WaitOnAddress` and `WakeByAddressSingle`, that are defined in `synchronization.lib`. Boost.Log does not reference this library during its (vcpkg-driven) build, but I'm told by the author of Boost.Log that when using Boost's own build system, it imports a dependency on it from Boost.Atomic. This apparently does not happen when vcpkg is in charge because vcpkg builds both libraries separately, just linking Log against Atomic. So Atomic knows that `synchronization.lib` exists and uses its features, but Log is not told that.

Patching a dependency on `synchronization.lib` into Boost.Log's `build/Jamfile.v2` results in a successful build and a working library, at least as far as I use it.

**Comment by shelllet** [comment]
![image](https://user-images.githubusercontent.com/16729017/153705176-96c70e90-1bdb-435a-bbf2-98b23562a132.png)

add `#define BOOST_USE_WINAPI_VERSION BOOST_WINAPI_VERSION_WIN7` seems to work.

I think this is just a temporary solution.

**Reply by take5v** [reply]
I locally made port overlay of boost-modular-build-helper where in CMakeLists.txt lines b2 options are added for the current WinAPI version (see https://www.boost.org/users/history/version_1_68_0.html):
```
if(WIN32)
    # We have to set _WIN32_WINNT for b2
    if(${CMAKE_SYSTEM_VERSION} EQUAL 10) # Windows 10
        list(APPEND B2_OPTIONS define=BOOST_USE_WINAPI_VERSION=0x0A00)
    elseif(${CMAKE_SYSTEM_VERSION} EQUAL 6.3) # Windows 8.1
        list(APPEND B2_OPTIONS define=BOOST_USE_WINAPI_VERSION=0x0603)
    elseif(${CMAKE_SYSTEM_VERSION} EQUAL 6.2) # Windows 8
        list(APPEND B2_OPTIONS define=BOOST_USE_WINAPI_VERSION=0x0602)
    elseif(${CMAKE_SYSTEM_VERSION} EQUAL 6.1) # Windows 7
        list(APPEND B2_OPTIONS define=BOOST_USE_WINAPI_VERSION=0x0601)
    elseif(${CMAKE_SYSTEM_VERSION} EQUAL 6.0) # Windows Vista
        list(APPEND B2_OPTIONS define=BOOST_USE_WINAPI_VERSION=0x0600)
    else() # Windows XP (5.1)
        list(APPEND B2_OPTIONS define=BOOST_USE_WINAPI_VERSION=0x0501)
    endif()
endif()
```

These lines must be added to all libs which use boost for a consistent abi.

Also, in boost-modular-build-helper port I found if condition to VCPKG_DETECTED_CMAKE_SYSTEM_NAME where WIN32_WINNT fixed strangely to 0x0A00. I'm not sure why this is needed.

**Comment by Neumann-A** [comment]
Hmm I see this problem in #25897 with `_WIN32_WINNT `set to `0x0A00`and run into the issue needing to link `synchronization.lib`

**Comment by nedelman** [comment]
I just encountered Boost.Log linker error as well, on Boost 1.79, with `_WIN32_WINNT=0x0A00` in my own project, using Visual Studio 2022 and a self-built version of Boost, as well as a pre-built version from Conan. Setting `_WIN32_WINNT=0x0601` resolves the linker issue in my project, but isn't suitable as a workaround in my case since other parts of my code depend on having a higher setting.

**Comment by rjayne-git** [comment]
I have this same boost_log_abi mismatch problem with vcpkg and no redefinition of _WIN32_WINNT.  I have a project that depends on boost.log, and I have vcpkg build them both with x64-windows triplet. So, doesn't that mean that boost.log is detecting the wrong _WIN32_WINNT?  Forcing _WIN32_WINNT via VCPKG_CXX_FLAGS seems like a work-around.  This should be re-promoted to an issue imo.
windows 10 pro
visual studio 2019
vcpkg release 2022.06.15.1
boost 1.79.0

**Comment by CandidCypher** [comment]
Recently ran into this same issue using boost libraries installed via Nuget (I wish I could use vcpkg but corp blocks use of a lot of command line tools). I am using VS 2022 and have installed the vc143 versions of 1.80.

**Comment by stephengtuggy** [comment]
I've tried most of the suggestions listed in this thread, but I'm still stuck with linker errors. Visual Studio 2022 on Windows 11. Latest vcpkg from master, as of yesterday.

It looks like the linker is trying to find the Boost libraries as .lib files, instead of .dll's. Why is it doing that? I have added `BOOST_ALL_DYN_LINK` to my compile definitions using `add_compile_definitions`. Does that setting not carry over from the compiler to the linker, or something?

**Reply by Neumann-A** [reply]
> It looks like the linker is trying to find the Boost libraries as .lib files

That is normal. Those are import libraries containing the symbols exported in the dll.

**Reply by chrullrich** [reply]
> I've tried most of the suggestions listed in this thread, but I'm still stuck with linker errors. Visual Studio 2022 on Windows 11. Latest vcpkg from master, as of yesterday.

https://github.com/chrullrich/vcpkg-local/commit/6160d5a37306a397b7c23b9b89169be7142b5729

**Reply by stephengtuggy** [reply]
@chrullrich Thanks. I'll take a look.

**Reply by Neumann-A** [reply]
ah if it is the synchro stuff. #27694 will fix it.

**Reply by stephengtuggy** [reply]
Still not working, even with your patches, @chrullrich . I was getting 40+ errors of the form:

```
error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl boost::log::v2_mt_nt62::sinks::text_file_backend::flush(void)"
```

Now I'm getting:

```
event.obj : error LNK2019: unresolved external symbol WaitOnAddress referenced in function "public: static unsigned int __cdecl boost::atomics::detail::wait_operations_windows<struct boost::atomics::detail::core_operations<4,0,0>,4>::wait(unsigned int const volatile &,unsigned int,enum boost::memory_order)" (?wait@?$wait_operations_windows@U?$core_operations@$03$0A@$0A@@detail@atomics@boost@@$03@detail@atomics@boost@@SAIAEDIIW4memory_order@4@@Z)
event.obj : error LNK2019: unresolved external symbol WakeByAddressSingle referenced in function "public: static void __cdecl boost::atomics::detail::wait_operations_windows<struct boost::atomics::detail::core_operations<4,0,0>,4>::notify_one(unsigned int volatile &)" (?notify_one@?$wait_operations_windows@U?$core_operations@$03$0A@$0A@@detail@atomics@boost@@$03@detail@atomics@boost@@SAXAECI@Z)
```

What's happening here? The .patch files do not seem to be taking effect for some reason.

**Reply by chrullrich** [reply]
Did you use one of the triplets I add in that commit, both for building Boost itself and your project?

![image](https://user-images.githubusercontent.com/4695907/201496434-52843184-733b-4292-bc35-f65eca27a9d9.png)

**Reply by stephengtuggy** [reply]
I set the environment variable `VCPKG_DEFAULT_TRIPLET` to `x64-windows-win10`. That ought to do it, right? And I saw `x64-windows-win10` showing up in the build output. Although triplet `x64-windows` by itself was also showing up, for reasons that I do not understand.

I'm using manifest mode (vcpkg.json). Does that make a difference?

**Reply by chrullrich** [reply]
Hm. I have never used manifest mode. Do I understand correctly that it rebuilds all dependencies for each project? I'll try it myself and let you know what happens.

I get the mixed triplets in what is now apparently "classic" mode, too. I don't think it matters.

**Reply by chrullrich** [reply]
The first random project I grabbed and dropped a vcpkg.json in built without such errors.

Random guess: I don't update the port versions, so could the build after adding the patches to the ports have reused either some previously cached binary packages, or just the buildtrees from before, i.e. not applied the patches to the Boost sources?

Also, can you try with @Neumann-A's fix instead?

**Reply by Neumann-A** [reply]
need to set VCPKG_HOST_TRIPLET in cmake or use --host-triplet to the same triplet as the normal triplet to not mix triplets.

**Reply by stephengtuggy** [reply]
@chrullrich That's entirely possible.

I'll try with `VCPKG_HOST_TRIPLET`. Thanks for the tip, @Neumann-A .

**Reply by stephengtuggy** [reply]
Yeah, looks like it's best to update the port versions whenever making changes to them.

I should note that I am also using CMakePresets.json, which is making life even more interesting haha.

At one point, I was able to get vcpkg to stop mixing triplets. But I still got the "unresolved external symbol" errors. Probably for the reason you mentioned, @chrullrich , that the port versions haven't been updated, so vcpkg doesn't recognize that it needs to re-download the boost source code.

**Reply by dg0yt** [reply]
> At one point, I was able to get vcpkg to stop mixing triplets. 

It is very easy: Build natively, i.e. with target triplet = host triplet.
But in general, it makes sense to use the same host triplet for all targets, in particular when they are not runnable on your host.

**Reply by chrullrich** [reply]
> Probably for the reason you mentioned, @chrullrich , that the port versions haven't been updated, so vcpkg doesn't recognize that it needs to re-download the boost source code.

It doesn't need to redownload, just reextract and patch. All that should be necessary is deleting the buildtrees. Since I assume you have done this, there must be something else wrong.

If you want to try again, keep hold of the build output. It mentions applied patches by name, so you can make sure they went in.

**Reply by stephengtuggy** [reply]
> > Probably for the reason you mentioned, @chrullrich , that the port versions haven't been updated, so vcpkg doesn't recognize that it needs to re-download the boost source code.
> 
> It doesn't need to redownload, just reextract and patch. All that should be necessary is deleting the buildtrees. Since I assume you have done this, there must be something else wrong.
> 
> If you want to try again, keep hold of the build output. It mentions applied patches by name, so you can make sure they went in.

I deleted the `boost-log` buildtree. Now that I think about it, maybe that wasn't sufficient.

For the time being, I'm just going back to Boost 1.77 using the baseline field in vcpkg.json. Not sure if or when I'll try again.

**Reply by stephengtuggy** [reply]
Tried again with @chrullrich 's code. Cleaned my build environment more thoroughly, and got a little further. Our .exe still doesn't link though. Now, it complains about a `freeglut` .obj file being built for a different ABI.

**Comment by ghost** [comment]
This problem is still present today, with boost 1.81, with the current vcpkg baseline. Has anyone found anything that works consistently?

**Reply by gaukem** [reply]
Agree. I get a error lnk2019 too when build project, what is linked with boost-log 1.81 (install from vcpkg) by VS2022 T_T

**Reply by stephengtuggy** [reply]
Using `"builtin-baseline": "b25cb779e5783531cc66562e224079b376055939"` in vcpkg.json works consistently, in my experience. Of course, you're downgrading Boost to version 1.77 when you do that.

Apart from that, you could try [this code](https://github.com/microsoft/vcpkg/pull/27694) to see if it works for you. Sorry I don't have a much better solution to offer at the moment. Hopefully a fix is found soon.

**Reply by ssh4net** [reply]
Still use 
`-DBOOST_USE_WINAPI_VERSION=BOOST_WINAPI_VERSION_WIN7 `
in c/c++ Comand line options.
Without this not possible to build a projects.

**Reply by ghost** [reply]
Sad. This entirely defeats the point of using a package manager. 

This is a major library, not a niche one, and on Microsoft's own platform. My company has already abandoned plans to adopt vcpkg over this. Too bad, it had potential. 

Thank you for the responses from participants in this thread. At this point it's just easier to compile boost myself though. 🤷‍♂️

**Reply by BillyONeal** [reply]
>Sad. This entirely defeats the point of using a package manager.

I'm not sure I understand, can you clarify? What is the experience you're expecting here? We can't know in advance what value of `BOOST_USE_WINAPI_VERSION` you want to use, but we could add mechanisms that make it easier for you to tell us.

>This is a major library, not a niche one, and on Microsoft's own platform. My company has already abandoned plans to adopt vcpkg over this. Too bad, it had potential.

Unfortunately the majorness of the library has no bearing on the situation that mutually exclusive alternatives that the whole world need to agree on create difficulty for customers. If you build boost yourself you're going to have to deal with the same consequences (of needing to set those macros consistently with how the Boost you're using is built).

**Reply by ghost** [reply]
I would guess that the assumption made by most users is that a library installed without special parameters (aside from using the correct triplet) will use the correct API for the toolchain and OS it is installed on. 

We tried vcpkg after watching a keynote given by a Microsoft employee at cppcon, and this was the workflow described. 

https://youtu.be/Ae9EePOIouU

I'm sure my answer could have been a bit tempered, and for that I apologize. But I do not see a way to make a reproducible multi-platform build using vcpkg this way. I never got my software to build on Windows. 

If there is a reproducible way to fix this without breaking the software's ability to be built cross-platform, I would be all ears, and it would probably benefit many who find this thread. 

I'm not a Windows developer by default, but I would prefer to offer my customers that choice. Vcpkg worked really great on Linux.

**Reply by BillyONeal** [reply]
>I would guess that the assumption made by most users is that a library installed without special parameters (aside from using the correct triplet) will use the correct API for the toolchain and OS it is installed on.

Our assumption is to be as compatible in terms of deployment as possible. If we made the defaults such that you couldn't deploy to Win7 people would complain. :(

>I'm sure my answer could have been a bit tempered, and for that I apologize.

No, it's fine. I'm asking mainly to understand not to call you olut.

>But I do not see a way to make a reproducible multi-platform build using vcpkg this way. I never got my software to build on Windows.

The question I'm trying to answer is: "is this something Boost.Log just *does*, and people who are using Boost.Log have to deal with it", vs. "is this something vcpkg itself did that we need to take action on". Everything in this thread to me looks like the former. boost requires a value for this setting, vcpkg picked one, and the one we picked doesn't happen to agree with what a user wanted. vcpkg exposes the same interface "upstream" does for the various libraries precisely because we want you to be able to use us in some circumstances and other providers in others. For example, vcpkg for Windows and apt or dnf for Linux are very common. But you're saying it's the latter (something vcpkg did), so I'm trying to understand where the gap is.

We can make it easier to get at the setting but we can't eliminate the need for everything everywhere to agree on the setting because this requirement comes from Boost.Log itself.

>If there is a reproducible way to fix this without breaking the software's ability to be built cross-platform, I would be all ears, and it would probably benefit many who find this thread.

I mean, we could create knobs that would allow adjustments of some of this stuff, like additional triplet variables.

**Reply by ghost** [reply]
Thank you for not taking offense. I was not at my best. 

Yes, if there is a place, really, anywhere, in vcpkg.json or in a triplet, or if there is already a way and several of us have overlooked it, that would be great.

Even guidance on how to customize the triplet ourselves (we can add it to our repo, so still reproducible) but I gave a good college try to figure it out, and failed. 

That can 100% be on me. If y'all are willing to do anything to help, I'll be grateful.

**Reply by stephengtuggy** [reply]
Best I can tell, Boost builds differently when built from within vcpkg. It doesn't follow the Boost standard build process. And there's a setting that is newly required as of Boost 1.78, that _is not set_ within vcpkg. (The setting is which version of Windows to make everything compatible with.)

Setting the Windows compatibility version to Win7 would be completely acceptable. The code would work. Everything would build and link fine. The problem is that the setting is not set, and users' projects default to a different value than Boost does, resulting in ABI incompatibility at link time (among other things).

And yes, the fact that Boost doesn't work from vcpkg is a pretty big deal.

**Reply by stephengtuggy** [reply]
Also: In our efforts to make things as compatible and flexible as possible, please, let's not neglect to make things "just work" out of the box. Not a lot of users are going to want to have to fiddle with triplet settings, if you ask me. I think that should be considered an advanced usage pattern.

**Reply by Neumann-A** [reply]
> Setting the Windows compatibility version to Win7 would be completely acceptable

Question: Why should vcpkg support Win7 when official support for it ended a long time ago and even extended support ended this year. 

From that point of view supporting Win7 seems to be like a advanced user setting which shouldn't be the default.  So if you want it add 
```
set(VCPKG_C_FLAGS -D_WIN32_WINNT=0x0601 -DWINVER=0x0601)
set(VCPKG_CXX_FLAGS -D_WIN32_WINNT=0x0601 -DWINVER=0x0601)
```
to your triplet. This also makes it consistent across all ports.

(The linkage of synchro.lib is another issue and fixed in #27694)

**Reply by chrullrich** [reply]
> Best I can tell, Boost builds differently when built from within vcpkg. It doesn't follow the Boost standard build process.

Yes, vcpkg basically builds each Boost library individually and maintains a list of dependencies, i.e. which library to link against which others. This approach loses valuable information about transitive dependencies, such as – in this case – that Boost.Log must link against `synchronization.lib` if `_WIN32_WINNT>=0x0602`. The Boost build system locates this information in Boost.Atomic and applies it when building Log. vcpkg knows to link Log against Atomic, it's just that that isn't always good enough.

> And there's a setting that is newly required as of Boost 1.78, that _is not set_ within vcpkg. (The setting is which version of Windows to make everything compatible with.)

Not quite. The change in 1.78 is the introduction of the `v2_mt_nt62` API in Boost.Log, which is the first one that uses features only found in Windows 8 and later. Prior versions of Log could build with `_WIN32_WINNT>=0x0602` because the highest supported required OS version was Windows 7.

> Setting the Windows compatibility version to Win7 would be completely acceptable. The code would work. Everything would build and link fine. The problem is that the setting is not set, and users' projects default to a different value than Boost does, resulting in ABI incompatibility at link time (among other things).

Building Boost with `_WIN32_WINNT=0x0602` (or 0x0A00) is trivial (you set `BOOST_USE_WINAPI_VERSION` as desired), _except_ for Boost.Log due to the missing dependency.

> And yes, the fact that Boost doesn't work from vcpkg is a pretty big deal.

Big enough that IMO, vcpkg should decide that the attempt to replace the Boost build system with its own has failed, throw it all away, and go back to using Boost's own. That will probably mean that Boost becomes an all-or-nothing selection again rather than allowing the user to select what libraries to build (or at least, that no changes to the selected library set are possible other than by removing the current set and building a new one), but that is a fairly cheap price to pay. If you want Log, that is.

(This would also get rid of the worst horror of them all, boost-uninstall.)

**Reply by Neumann-A** [reply]
>  That will probably mean that Boost becomes an all-or-nothing selection again

then people will start complaining that their header only boost dependency takes too long to build ;)

> (This would also get rid of the worst horror of them all, boost-uninstall.)

all boost ports should have gone through boost-build instead of having header only boost ports treated specially. 

>  that Boost.Log must link against synchronization.lib

That is fixed in #27694

**Reply by chrullrich** [reply]
> > that Boost.Log must link against synchronization.lib
> 
> That is fixed in #27694

I know, and have mentioned it in the past, but when will it be merged?

**Reply by stephengtuggy** [reply]
Okay, so I figured something out. At least enough for my immediate needs.

I performed the following steps:

1. Created my own [fork](https://github.com/stephengtuggy/vcpkg-local) of @chrullrich 's [fork](https://github.com/chrullrich/vcpkg-local) of vcpkg
2. In `triplets/x64-windows.cmake`, set `VCPKG_C_FLAGS` and `VCPKG_CXX_FLAGS` to include `/DBOOST_USE_WINAPI_VERSION=0x0A00 /D_WIN32_WINNT=0x0A00 /DWINVER=0x0A00`. Set the same flags in `triplets/x64-windows-win10.cmake`
3. In my Windows environment variables, set `VCPKG_ROOT` to the local working copy of my vcpkg fork. Also set `VCPKG_DEFAULT_HOST_TRIPLET` and `VCPKG_DEFAULT_TRIPLET` to `x64-windows-win10`. (Setting them to `x64-windows` would probably also work.)
4. Made sure the build trees were clean by simply deleting the `buildtrees` directory under the vcpkg local working copy
5. In my project, in `CMakeLists.txt`, added the following block, repeated for each target executable or library:

```
target_compile_definitions(vegastrike-OPcollide PUBLIC "BOOST_ALL_DYN_LINK" "$<$<CONFIG:Debug>:BOOST_DEBUG_PYTHON>" "$<$<CONFIG:Debug>:Py_DEBUG>")
target_compile_definitions(vegastrike-OPcollide PUBLIC BOOST_USE_WINAPI_VERSION=0x0A00)
target_compile_definitions(vegastrike-OPcollide PUBLIC _WIN32_WINNT=0x0A00)
target_compile_definitions(vegastrike-OPcollide PUBLIC WINVER=0x0A00)
```

6. In `vcpkg.json`, took out the `builtin-baseline` setting, along with all version overrides for individual dependencies.
7. Added the following line at the top of every C++ source file that involves Python-related code:

```
#define PY_SSIZE_T_CLEAN
```

There were a few other changes I made, but I think those were the key differences. With these changes, I was able to:

- Use Boost.Log
- Use Boost.Python
- Upgrade Boost to the latest version available from vcpkg, 1.81
- Get embedded Python code to work in Debug configuration on Windows, which it didn't before. 

Overall, I am very happy with the outcome. You can find a minimal reproducible example here: https://github.com/stephengtuggy/boost_log_python_poc

**Reply by stephengtuggy** [reply]
And here is the relevant Pull Request on the original project: https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/776 (Sorry, it's a rather large set of changes, not all of which are directly related to this issue.)

**Comment by ssh4net** [comment]
Is it possible to disable VCPKG to use Boost from the port, but use "external" from boost repo to build packages?

**Reply by Neumann-A** [reply]
needs an empty overlay for every boost port.
